### DISCUSSION TITLE
ultimate version crash....

### DISCUSSION THREAD

**Discussion by amtrak115** [perpetualKid/FreeTrainSimulator#67] [General]
went to the Release , and downloaded 1.3.3-rc.1.zip.  installed, then ran,  did the update to 1.3.6 (I believe) then decided to run a quick test.  started a train, 3 loco's 56 loaded coal hoppers, 3 helper locos, 56 loaded coal hoppers, leaving Raton yard (Raton Pass III).  just before leaving yard stop train, then using "Train OP's button" on RailDriver, scrolled the window over until 3 helper engines appear, Use mouse to "uncouple" coal cars from rear of helpers.  OR Crashes.  I assume since your keeping this version separate from other versions....this was proper place to inform you.  If not let me know.  

Barry
[OpenRailsLog.txt](https://github.com/perpetualKid/ORTS-MG/files/6281269/OpenRailsLog.txt)

**Comment by perpetualKid** [comment]
Thanks Barry. I've converted that into an issue on the bugtracker directly, feel free for similar issues to do the same. 
However we're still sorting out the flow here, so nothing wrong posting into discussions - as long issues are getting reported at all so they can be tracked and worked on.

just an FYI since you're using RailDriver - as there is currently some bigger work ongoing on the overall input handling (a whole new (and much improved) system which we first implemented and tested in the new Trackviewer, RailDriver support is currently broken after 1.3.6 version (so anything 1.3.7, which is currently available only in the dev/ci update channel). This is not a bug but work in progress, and will be updated once the keyboard&mouse input implementation is done. But this should not crash the program, only the button on RD currently do not work in any 1.3.7.* version.

I do think though the bug above is something different and not related, will need to look into it.

**Comment by perpetualKid** [comment]
this should be fixed with https://github.com/perpetualKid/ORTS-MG/commit/01327e3d81cbfa7fe4087de372ab0e229a0d5b69 but only packaged yet in the CI updater channel

**Reply by amtrak115** [reply]
dumb question.....  What is the CI channel?

**Reply by perpetualKid** [reply]
In the options menu under the Updater tab, there are differnt updater channels. This is somewhat similar as in OR, but not the same. CI stands for **C**ontinuous **I**ntegration, basically everytime we check in some code changes, a build process is triggered and creates a new version. However in the CI channel, different code changes may happen from various branches, leading sometimes to unpredictable result. (this requires some work to streamline... [already planned]( https://github.com/perpetualKid/ORTS-MG/projects/7)). RC (Release Candidate) and Release channel are (more) stable builds, but manually triggered at certain stages only.

![image](https://user-images.githubusercontent.com/7158903/114232754-af3df600-997c-11eb-8804-26c37eb4f0c3.png)

**Comment by perpetualKid** [comment]
@amtrak115 how are things going? Just wanted to let you know that RailDriver support is back!
It's available in the Monogame branch for now , and also in binary form from version #1.3.7-ci.36 

There are some minor improvements still work in progress, but in generall all the RailDriver functionality is restored/reimplemented based on our new input framework. 
If you experience any issues, let us know.

**Reply by amtrak115** [reply]
making some progress.  I was not able to rebuild the entire OR on my system.  it threw an error or a number of errors.  will have to look at that again.  was able to rebuild the trackviewer proj without any issues however.  will download the monogame source again to regain RailCommand.  i have a number of questions that I've come up with...but am trying to figure out through the code.  Also was working on an Activity that I've been working on..  playing with Location events and such...anyway...I'll probably be sending questions over your way in the next couple of days...I would still like to investigate more.

**Reply by amtrak115** [reply]
did an update.  it pulled in 1.3.7-ci.39.  my raildriver does not respond.  tried to callibrate, not sure if that made a difference as I still have nothing with raildriver.

**Reply by perpetualKid** [reply]
did you activate the raildriver when the game starts? By default, it will not be active, and you have to activate from Keyboard (using \` key in default keyboard mapping). When not active, the display on the rd should should three dashes (---), but when active, will show the actual train speed. Does it activate the display when the game starts, or is there something else?
Maybe I should set the RD to active by default when the game starts, or make it configurable, and also put some trace messages in the log when/if the RailDriver is found and enabled....

as for the code, it should build just fine, first time it may need to restore a couple packages so online connectivity is required. If you get stuck, post the errors or some log so we can try to figure out.

**Reply by amtrak115** [reply]
Got RailDriver working....there appears to be conflicting definitions for "Activate raildriver"...in Keyboard mapping use ` to activate...that works, in RailDriver definitions,  button 14 activates except in this case it deactivates Raildriver.  have to use ` to reactivate.  in version OR-NewYear-MG version 96 (released a couple of days ago)  both ` and button 14 on RD will activate or deactivate.  Button 14 is labeled "RailDriver Run/Stop" from "standard MSTS button layout from RailDriver Website".  I know I'm getting some errors on the rebuild about a file/files's  "having the mark of the web" on them.  I'll be looking at that in a little bit.  Will let you know.

**Reply by perpetualKid** [reply]
I think in this new input framework unintentionally the option got removed to activate the RailDriver from itself. When it's not active now, its basically like shutoff, and won't react ("awake") on any input. Deactivate though works...
Thats why I above thought to optionally make RD configurably active at game startup.
Or (less prefered) implement some workaround to restore the old behavior - effectively means the RD would be (hardware) active all the time but remain passive until "activated". Thoughts?

**Reply by amtrak115** [reply]
I think making RD configurably active is the way to go.  I use RD all the time, having to work a locomotive just from Keyboard is painful in my opinion.

**Reply by amtrak115** [reply]
with respect to the errors I'm getting when rebuilding open rails.....I was getting some errors related to .resx file not being unblocked.  have corrected that.  I'm not asking you to walk me through fixing all of these, just point me in the right direction so I can learn more about the dev-environment and the way the code works....

Barry
![Capture1](https://user-images.githubusercontent.com/22547728/114953271-a4241380-9e1d-11eb-80df-c0de8d067ad7.JPG)

**Reply by perpetualKid** [reply]
Somehow I think your codebase is not in a consistent state - files seem to belong to different code versions, hence failing to build.   

How did you download the code? I.e. it strikes me as odd, that you need to unblock resx files. Are you using git client or commandline? The integrated tool in Visual Studio is not much intuitive, and unless you're fond in CLI, I'd recommend [GitHub Desktop](https://desktop.github.com/). It's fairly straight forward to setup, and you can easily change between branches and sync the latest updates.

**Reply by amtrak115** [reply]
What I did, mainly to not screw up anything your working on.  Went the Monograme version and clicked on Download.  I have not looked at the GitHub Desktop, Bu I will do that.
Barry

**Reply by perpetualKid** [reply]
no worries, it's all read-only until we give you some dedicated permissions ;)
Definitely recommend using some version-control tool, even just for your private work, i.e where you can fork our repo once and keep working in your own copy, even privately if you want. 
If you're interested in some more hands-on details, I suggest we change over to our new Gitter channel [![Join the chat at https://gitter.im/ORTS-MG/community](https://badges.gitter.im/ORTS-MG/community.svg)](https://gitter.im/ORTS-MG/community?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) which is more of a chat-type environment.
